### PR TITLE
Use Puppet SSL by default

### PR DESCRIPTION
**Pull request by treydock**
# Pull Request Checklist

## Description

By default the backend will be configured to use SSL and agents will communicate using SSL.  The certificates and CA used are from Puppet.

This adds several parameters.

**sensu**:
* ssl_dir
* ssl_dir_purge
* use_ssl
* ssl_ca_source
* ssl_add_ca_trust

**sensu::backend**:
* url_host - default changed to `certname`
* use_ssl - moved to `sensu` class
* ssl_cert_source
* ssl_key_source

**sensu::agent**:
* backends - used to generate `backend-url` config option

Added dependency on `puppet/trusted_ca` module.

## Motivation and Context

Puppet already has SSL certificates and would be good to default sensu-go to use a secure configuration.

## How Has This Been Tested?

Unit and beaker

## General

- [x] Update `README.md` with any necessary configuration snippets

- [x] New parameters are documented

- [x] New parameters have tests

- [x] Tests pass - `bundle exec rake validate lint spec`

### CODE REVIEW

**Comment by treydock** [comment]
Test failures I can't tell are issue with our dependencies or something broken at rubygems.org.  This PR did not modify `Gemfile` so thinking something is wrong with rubygems.org or travis-ci.

**Comment by treydock** [comment]
I disabled SSL for vagrant environment.  I wasn't sure if we wanted to add puppetserver logic to Vagrant test environment.

**Comment by treydock** [comment]
I wanted to use `camptocamp/openssl` for the `openssl` package but that module and `trusted_ca` module have conflicting resources.  I will try to fix `trusted_ca` module so we can use `openssl` module and not define the package in this module if that sounds like something we should do.

Also need to add the Puppet config facts in this PR to stdlib module.  Right now stdlib looks at Puppet config settings for a couple facts but I think their implementation doesn't work if `puppet.conf` doesn't have the value.  The method used in this PR will return puppet's default values if they are not defined in `puppet.conf`.

File: lib/facter/sensu_puppet_facts.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
+require 'puppet'
+
+module PuppetFacts
```
Comment (ghoneycutt) [inline lib/facter/sensu_puppet_facts.rb]: does `PuppetFacts` come from puppet or from a gem?

Comment (treydock) [reply lib/facter/sensu_puppet_facts.rb]: It's defined in that file, I should have probably used `SensuPuppetFacts` to match file name.

**Comment by ghoneycutt** [comment]
If we're going to SSL by default, then it should be that way in Vagrant, too.

File: lib/facter/sensu_puppet_facts.rb (hunk 1 above)
Comment (ghoneycutt) [reply lib/facter/sensu_puppet_facts.rb]: Aha, let's do that to make it more apparent.

**Comment by treydock** [comment]
The failures in tests appear to be specific to puppetserver which is odd given puppetserver is running on CentOS 7 for all OSes in the test matrix, so would expect it to fail for all and not just some.  I'm not having much luck reproducing the failures when running beaker locally.  Will investigate as just appears like puppetserver is getting timed out when starting.

File: Vagrantfile [hunk 2]
```diff
@@ -30,7 +30,8 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
   config.vm.synced_folder ".", "/vagrant", type: "virtualbox"
 
   config.vm.provider :virtualbox do |vb|
-    vb.customize ["modifyvm", :id, "--memory", "512"]
+    vb.customize ["modifyvm", :id, "--memory", "1512"]
+    vb.cpus = 2
```
Comment (ghoneycutt) [inline Vagrantfile]: Why add CPU's?

File: manifests/ssl.pp [hunk 3]
```diff
@@ -0,0 +1,37 @@
+# @summary Private class to manage sensu SSL resources
+#
```
Comment (ghoneycutt) [inline manifests/ssl.pp]: I think it still needs `@api private`

File: manifests/repo.pp [hunk 4]
```diff
@@ -1,4 +1,5 @@
-# private class
+# @summary Private class to manage sensu repository resources
+#
```
Comment (ghoneycutt) [inline manifests/repo.pp]: I think it still needs `@api private`

File: spec/acceptance/00_backend_spec.rb [hunk 5]
```diff
@@ -5,7 +5,8 @@
   context 'default' do
     it 'should work without errors' do
       pp = <<-EOS
-      include ::sensu::backend
+      class { '::sensu': }
+      class { '::sensu::backend': }
```
Comment (ghoneycutt) [inline spec/acceptance/00_backend_spec.rb]: curious why the change to resource syntax for the inclusion of the class.

File: types/backend_url.pp [hunk 6]
```diff
@@ -0,0 +1,5 @@
+type Sensu::Backend_URL = Variant[
```
Comment (ghoneycutt) [inline types/backend_url.pp]: Check out this PR to stdlib as an example for writing spec tests for type aliases

https://github.com/puppetlabs/puppetlabs-stdlib/commit/6dad21e292436ac5b7d8113e81e572a0efc66782

**Comment by ghoneycutt** [comment]
We should also have tests that show this works with puppet apply when setting use_ssl to false.

File: Vagrantfile (hunk 2 above)
Comment (treydock) [reply Vagrantfile]: Had a few occurrences of startup timing out when testing and figured more CPUs would help given how resource intensive puppetserver can be.

File: spec/acceptance/00_backend_spec.rb (hunk 5 above)
Comment (treydock) [reply spec/acceptance/00_backend_spec.rb]: I wanted to verify the README style of examples would work.  If `class {}` works then we know `include` will work.

File: spec/acceptance/03_no_ssl_spec.rb [hunk 7]
```diff
@@ -0,0 +1,62 @@
+require 'spec_helper_acceptance'
+
+describe 'sensu without SSL', unless: RSpec.configuration.sensu_cluster do
```
Comment (treydock) [inline spec/acceptance/03_no_ssl_spec.rb]: @ghoneycutt This is the tests where `use_ssl => false` is done using method documented in README.

**Comment by treydock** [comment]
@ghoneycutt the part listed twice you mean the `sed` commands?  They are actually different.  One is `Xms` and the other is `Xmx`.

File: Vagrantfile (hunk 2 above)
Comment (ghoneycutt) [reply Vagrantfile]: This can slow things down with virtualization as now there needs to be two free cpu's available at the same time. If your system only has one at that moment, then the hypervisor waits until both are free.

Comment (treydock) [reply Vagrantfile]: Removed cpus change.

File: manifests/repo.pp (hunk 4 above)
Comment (treydock) [reply manifests/repo.pp]: Added `@api private` back

File: types/backend_url.pp (hunk 6 above)
Comment (treydock) [reply types/backend_url.pp]: Added unit tests based on the stdlib example.

File: spec/type_aliases/backend_url_spec.rb [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    ].each do |value|
+      describe value.inspect do
+        it { is_expected.to allow_value(value) }
+      end
+    end
+  end
+
+  context 'disallow invalid values' do
+    [
+      'http://localhost:8081',
```
Comment (ghoneycutt) [inline spec/type_aliases/backend_url_spec.rb]: you can't have a HTTP backend?

File: tests/provision_basic_el.sh [hunk 9]
```diff
@@ -31,4 +31,9 @@ puppet resource file /etc/puppetlabs/code/environments/production/modules/sensu
 
 # setup module dependencies
 puppet module install puppetlabs/stdlib --version 4.25.1
+puppet module install puppet/trusted_ca --version 2.0.0
+
+puppet resource host puppet ensure=present ip=192.168.52.10
+puppet resource host sensu-backend.example.com ensure=present ip=192.168.52.10
+puppet agent --test --tags no_such_tag
```
Comment (ghoneycutt) [inline tests/provision_basic_el.sh]: could you use `-t` instead of `--test`. Most people learn with `-t` as the name 'test' is vary misleading.

File: spec/type_aliases/backend_url_spec.rb (hunk 8 above)
Comment (treydock) [reply spec/type_aliases/backend_url_spec.rb]: Not that I'm aware of, docs say use `ws://` or `wss://` (ssl).

File: tests/provision_basic_el.sh (hunk 9 above)
Comment (treydock) [reply tests/provision_basic_el.sh]: Changed.

File: spec/type_aliases/backend_url_spec.rb (hunk 8 above)
Comment (treydock) [reply spec/type_aliases/backend_url_spec.rb]: Confirmed with vagrant, have to use `ws://` or `wss://`, tried using `https://`:

```
Jan  3 01:02:51 el7-agent sensu-agent: {"component":"agent","error":"malformed ws or wss URL","level":"error","msg":"reconnection attempt failed","time":"2019-01-03T01:02:51Z"}
```

**Comment by treydock** [comment]
I'll squash commits down once I get acceptance tests working.  I'm not liking how puppetserver is behaving and causing only some of the tests to fail.  It makes no sense because all OSes in matrix use exact same docker image (centos:7) for puppetserver and only the same 3 seem to fail.  A really ugly fallback solution could be to generate the puppet certs with Vagrant, store them in the repo and copy into test docker containers in same location that Puppet would use.

**Comment by treydock** [comment]
Updated to store Puppet SSL certs in `tests/ssl` directory.  The backend cert has a very long Subject Alt Name that allows it to work with both Vagrant and docker.  I created copies of the private key and certs for backend so that the default host certname would pick up the correct certificate that has SAN.

I also changed how acceptance test roles are looked up because ran into issues where only one of the 3 backend cluster nodes for cluster tests had `sensu_backend` role and this was incorrect and previous role lookup only allowed role to exist on one node in a test which didn't seem right when cluster had 3 that needed same role.

**Comment by treydock** [comment]
@ghoneycutt tests are now passing, this is good to merge if everything looks good.

File: spec/unit/facter/sensu_puppet_facts_spec.rb [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+    Puppet[:codedir] = '/tmp'
+    subject.add_facts
+  end
+  after(:each) do
+    Facter.clear
+    Facter.clear_messages
+  end
+
+  it 'should have puppet_hostcert defined' do
+    Puppet[:hostcert] = '/dne/host.pem'
```
Comment (ghoneycutt) [inline spec/unit/facter/sensu_puppet_facts_spec.rb]: what's dne?

File: types/backend_url.pp (hunk 6 above)
Comment (ghoneycutt) [inline types/backend_url.pp]: It can be useful to comment regex's with what they should match, such as  the following. This makes it easy to grok what's happening.

`Sensu Backend URL's use ws:// or wss:// or can be specified with the host only and no protocol. They must always have the port specified.`

File: spec/unit/facter/sensu_puppet_facts_spec.rb (hunk 10 above)
Comment (treydock) [reply spec/unit/facter/sensu_puppet_facts_spec.rb]: Does Not Exist.  Just passing in something that doesn't exist for purposes of unit testing.

File: types/backend_url.pp (hunk 6 above)
Comment (treydock) [reply types/backend_url.pp]: So `ws://` or `wss://` are actually always required but made optional because they will be automatically added by `sensu::agent` class based on value of `use_ssl`.  I am making note of that in comment.
